File: analysis/licitacao_gaps.py
from typing import Any

import pandas as pd
from sqlalchemy import text

from analysis.constants import dispensation_threshold

SAUDE_EMPRESA = "2"


def counts_by_year(conn: Any, years: list[int], empresa_ids: list[str] | None = None) -> dict[int, int]:
    """Retorna {ano: contagem_acima_do_limite} usando o limite aplicável por contrato."""
    placeholders = ", ".join(str(y) for y in years)
    empresa_clause = "AND empresa_id = ANY(:empresas)" if empresa_ids else ""
    params: dict = {}
    if empresa_ids:
        params["empresas"] = empresa_ids
    df = pd.read_sql_query(
        text(
            f"SELECT ano, licitacao_numero, valor_contrato, numero_obra, tipo_obra, objeto"
            f" FROM fct_contratos WHERE ano IN ({placeholders}) {empresa_clause}"
        ),
        conn,
        params=params,
    )
    df = df[df["licitacao_numero"].fillna("").str.strip() == ""].copy()
    df["valor_num"] = pd.to_numeric(df["valor_contrato"].astype(str).str.replace(",", "."), errors="coerce").fillna(0)
    df["threshold"] = df.apply(
        lambda r: dispensation_threshold(r.get("numero_obra"), r.get("tipo_obra"), r.get("objeto")), axis=1
    )
    df["acima_limite"] = df["valor_num"] > df["threshold"]
    counts = df.groupby("ano")["acima_limite"].sum().astype(int)
    return {y: int(counts.get(y, 0)) for y in years}


def totals_sem_licitacao_por_ano(conn: Any, years: list[int], empresa_ids: list[str] | None = None) -> dict[int, int]:
    """Retorna {ano: total_contratos_sem_licitacao} para todos os valores (acima e abaixo do limite)."""
    placeholders = ", ".join(str(y) for y in years)
    empresa_clause = "AND empresa_id = ANY(:empresas)" if empresa_ids else ""
    params: dict = {}
    if empresa_ids:
        params["empresas"] = empresa_ids
    df = pd.read_sql_query(
        text(f"SELECT ano, licitacao_numero FROM fct_contratos WHERE ano IN ({placeholders}) {empresa_clause}"),
        conn,
        params=params,
    )
    df = df[df["licitacao_numero"].fillna("").str.strip() == ""]
    counts = df.groupby("ano").size()
    return {y: int(counts.get(y, 0)) for y in years}


def run(conn: Any, year: int, empresa_ids: list[str] | None = None) -> pd.DataFrame:
    empresa_clause = "AND empresa_id = ANY(:empresas)" if empresa_ids else ""
    params: dict = {"ano": year}
    if empresa_ids:
        params["empresas"] = empresa_ids
    df = pd.read_sql_query(
        text(
            "SELECT ano, empresa_id AS empresa, contrato_numero AS numero, fornecedor_nome AS fornecedor, objeto, valor_contrato, licitacao_numero, mes,"
            " numero_obra, tipo_obra, modalidade, fundlegal"
            f" FROM fct_contratos WHERE ano = :ano {empresa_clause}"
        ),
        conn,
        params=params,
    )
    df = df[df["licitacao_numero"].fillna("").str.strip() == ""].copy()
    df["valor_num"] = pd.to_numeric(df["valor_contrato"].astype(str).str.replace(",", "."), errors="coerce").fillna(0)
    df["threshold"] = df.apply(
        lambda r: dispensation_threshold(r.get("numero_obra"), r.get("tipo_obra"), r.get("objeto")), axis=1
    )
    df["acima_limite"] = df["valor_num"] > df["threshold"]
    df["orgao_saude"] = df["empresa"] == SAUDE_EMPRESA
    df["periodo"] = df["mes"].astype(str).str.zfill(2) + "/" + df["ano"].astype(str)
    return df.drop(columns=["valor_num"]).rename(columns={"threshold": "limite_dispensa"})


def filter_above_limit(df: pd.DataFrame) -> pd.DataFrame:
    return df[df["acima_limite"]]


def filter_above_limit_health(df: pd.DataFrame) -> pd.DataFrame:
    return df[df["acima_limite"] & df["orgao_saude"]]

File: packages/db/src/queries/licitacao_gaps.ts
import { sql } from "kysely";
import { db } from "../client";
import { dispensationThreshold } from "../constants";

export const SAUDE_EMPRESA = "2";

export interface ContratoSemLicitacao {
  ano: number;
  empresa: string;
  numero: string;
  fornecedor: string;
  objeto: string;
  valorContrato: string;
  licitacaoNumero: string;
  mes: number;
  numeroObra: string | null;
  tipoObra: string | null;
  modalidade: string | null;
  fundlegal: string | null;
  limiteDispensa: number;
  acimaLimite: boolean;
  orgaoSaude: boolean;
  periodo: string;
  isentoLegalmente?: boolean;
}

export async function countsByYear(
  years: number[],
  empresaIds?: string[] | null,
): Promise<Record<number, number>> {
  const result: Record<number, number> = {};
  for (const y of years) result[y] = 0;

  try {
    let q = sql`
      SELECT ano, licitacao_numero, valor_contrato, numero_obra, tipo_obra, objeto
      FROM fct_contratos
      WHERE ano = ANY(${years})
    `;
    if (empresaIds && empresaIds.length > 0) {
      q = sql`${q} AND empresa_id = ANY(${empresaIds})`;
    }
    const res = await q.execute(db);
    const rows = (res.rows as any[]) || [];

    for (const r of rows) {
      const licNum = String(r.licitacao_numero ?? "").trim();
      if (licNum !== "") continue;

      const y = Number(r.ano);
      const val =
        parseFloat(String(r.valor_contrato ?? "0").replace(",", ".")) || 0;
      const th = dispensationThreshold(r.numero_obra, r.tipo_obra, r.objeto);

      if (val > th) {
        result[y] = (result[y] || 0) + 1;
      }
    }
  } catch {}

  return result;
}

export async function totalsSemLicitacaoPorAno(
  years: number[],
  empresaIds?: string[] | null,
): Promise<Record<number, number>> {
  const result: Record<number, number> = {};
  for (const y of years) result[y] = 0;

  try {
    let q = sql`
      SELECT ano, licitacao_numero
      FROM fct_contratos
      WHERE ano = ANY(${years})
    `;
    if (empresaIds && empresaIds.length > 0) {
      q = sql`${q} AND empresa_id = ANY(${empresaIds})`;
    }
    const res = await q.execute(db);
    for (const r of (res.rows as any[]) || []) {
      const licNum = String(r.licitacao_numero ?? "").trim();
      if (licNum === "") {
        const y = Number(r.ano);
        result[y] = (result[y] || 0) + 1;
      }
    }
  } catch {}

  return result;
}

export async function getLicitacaoGaps(
  year: number,
  empresaIds?: string[] | null,
): Promise<ContratoSemLicitacao[]> {
  try {
    let q = sql`
      SELECT ano, empresa_id AS empresa, contrato_numero AS numero, fornecedor_nome AS fornecedor, objeto, valor_contrato, licitacao_numero, mes, numero_obra, tipo_obra, modalidade, fundlegal
      FROM fct_contratos
      WHERE ano = ${year}
    `;
    if (empresaIds && empresaIds.length > 0) {
      q = sql`${q} AND empresa_id = ANY(${empresaIds})`;
    }
    const res = await q.execute(db);
    const list: ContratoSemLicitacao[] = [];

    for (const r of (res.rows as any[]) || []) {
      const licNum = String(r.licitacao_numero ?? "").trim();
      if (licNum !== "") continue;

      const emp = String(r.empresa ?? "");
      const val =
        parseFloat(String(r.valor_contrato ?? "0").replace(",", ".")) || 0;
      const th = dispensationThreshold(r.numero_obra, r.tipo_obra, r.objeto);
      const mStr = String(r.mes ?? "").padStart(2, "0");

      list.push({
        ano: Number(r.ano),
        empresa: emp,
        numero: String(r.numero ?? ""),
        fornecedor: String(r.fornecedor ?? ""),
        objeto: String(r.objeto ?? ""),
        valorContrato: String(r.valor_contrato ?? "0"),
        licitacaoNumero: licNum,
        mes: Number(r.mes),
        numeroObra: r.numero_obra ?? null,
        tipoObra: r.tipo_obra ?? null,
        modalidade: r.modalidade ?? null,
        fundlegal: r.fundlegal ?? null,
        limiteDispensa: th,
        acimaLimite: val > th,
        orgaoSaude: emp === SAUDE_EMPRESA,
        periodo: `${mStr}/${r.ano}`,
      });
    }

    return list;
  } catch {
    return [];
  }
}

export interface ItemDistribucaoModalidade {
  modalidade: string;
  valorTotal: number;
  quantidade: number;
  pctValor: number;
}

export async function getDistribucaoModalidades(
  year: number,
  empresaIds?: string[] | null,
): Promise<ItemDistribucaoModalidade[]> {
  try {
    let q = sql`
      SELECT
        COALESCE(NULLIF(TRIM(modalidade), ''), 'Outros') AS modalidade,
        COUNT(*)::int AS quantidade,
        SUM(COALESCE(valor_contrato, 0))::numeric AS valor_total
      FROM fct_contratos
      WHERE ano = ${year}
    `;
    if (empresaIds && empresaIds.length > 0) {
      q = sql`${q} AND empresa_id = ANY(${empresaIds})`;
    }
    q = sql`${q} GROUP BY 1 ORDER BY valor_total DESC`;

    const res = await q.execute(db);
    const rows = (res.rows as any[]) || [];

    const grandTotal = rows.reduce(
      (acc, r) => acc + (parseFloat(String(r.valor_total ?? "0")) || 0),
      0,
    );

    return rows.map((r) => {
      const val = parseFloat(String(r.valor_total ?? "0")) || 0;
      return {
        modalidade: String(r.modalidade ?? "Outros"),
        valorTotal: val,
        quantidade: Number(r.quantidade ?? 0),
        pctValor: grandTotal > 0 ? (val / grandTotal) * 100 : 0,
      };
    });
  } catch {
    return [];
  }
}
